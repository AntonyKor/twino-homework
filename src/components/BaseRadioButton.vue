<template>
  <span>
    <input
      v-bind="$attrs"
      :checked="modelValue === value"
      :value="value"
      @change="$emit('change', value)"
      type="radio"
      class="radio"
    />
    <span class="checkmark"></span>
  </span>
</template>

<script>
export default {
  inheritAttrs: false,
  model: {
    prop: "modelValue",
    event: "change",
  },
  props: {
    modelValue: { type: [String, Number], default: "" },
    value: { type: [String, Number], required: true },
  },
};
</script>

<style lang="scss" scoped>
.radio {
  display: none;

  &:checked ~ .checkmark {
    border: 5px solid #2e50b6;
  }
}

.checkmark {
  content: "";
  display: inline-block;
  box-sizing: border-box;
  height: 24px;
  width: 24px;
  border: 1px solid #9e9fa2;
  border-radius: 50%;
  margin-right: 12px;
}
</style>
