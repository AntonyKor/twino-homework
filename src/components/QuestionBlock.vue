<template>
  <div>
    <h3>{{ question.question }}</h3>
    <p v-for="(option, index) in question.answers" :key="option">
      <label>
        <BaseRadioButton v-model="answer" :value="index" name="answer" />
        {{ option }}
      </label>
    </p>
  </div>
</template>

<script>
import BaseRadioButton from "./BaseRadioButton.vue";

export default {
  components: {
    BaseRadioButton,
  },

  props: {
    question: {
      type: Object,
      required: true,
    },
  },

  computed: {
    answer: {
      get() {
        return this.question.answer;
      },
      set(answer) {
        this.$emit("set-answer", answer);
      },
    },
  },
};
</script>

<style lang="scss" scoped>
label {
  display: flex;
  align-items: center;
}

div {
  margin-bottom: 30px;
}
</style>
