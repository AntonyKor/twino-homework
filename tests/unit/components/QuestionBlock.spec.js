import { shallowMount } from "@vue/test-utils";
import QuestionBlock from "@/components/QuestionBlock.vue";
import BaseRadioButton from "@/components/BaseRadioButton.vue";

describe("QuestionBlock", () => {
  const question = {
    question: "question text",
    answers: ["a1", "a2", "a3"],
    answer: 1,
  };

  it("should match snapshot", () => {
    const wrapper = shallowMount(QuestionBlock, {
      propsData: { question },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it("should render answers", () => {
    const wrapper = shallowMount(QuestionBlock, {
      propsData: { question },
    });

    expect.hasAssertions();
    const labels = wrapper.findAll("label");
    question.answers.forEach((text, i) =>
      expect(labels.at(i).text()).toEqual(text)
    );
  });

  it("should pass current answer to modelValue to BaseRadioButton", () => {
    const wrapper = shallowMount(QuestionBlock, {
      propsData: { question },
    });

    expect.hasAssertions();
    const buttons = wrapper.findAllComponents(BaseRadioButton);
    question.answers.forEach((_, i) =>
      expect(buttons.at(i).props("modelValue")).toEqual(question.answer)
    );
  });

  it("should pass index to value to BaseRadioButton", () => {
    const wrapper = shallowMount(QuestionBlock, {
      propsData: { question },
    });

    expect.hasAssertions();
    const buttons = wrapper.findAllComponents(BaseRadioButton);
    question.answers.forEach((_, i) =>
      expect(buttons.at(i).props("value")).toEqual(i)
    );
  });

  it("should pass emit set-answer on change event on BaseRadioButton", () => {
    const wrapper = shallowMount(QuestionBlock, {
      propsData: { question },
    });

    wrapper.findAllComponents(BaseRadioButton).at(2).vm.$emit("change", 2);

    expect(wrapper.emitted("set-answer")).toHaveLength(1);
    expect(wrapper.emitted("set-answer")[0]).toEqual([2]);
  });
});
