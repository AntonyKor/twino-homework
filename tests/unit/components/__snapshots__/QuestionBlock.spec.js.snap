// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionBlock should match snapshot 1`] = `
<div>
  <h3>
    question text
  </h3>
   
  <p>
    <label>
      <baseradiobutton-stub
        modelvalue="1"
        name="answer"
        value="0"
      />
      
      a1
    
    </label>
  </p>
  <p>
    <label>
      <baseradiobutton-stub
        modelvalue="1"
        name="answer"
        value="1"
      />
      
      a2
    
    </label>
  </p>
  <p>
    <label>
      <baseradiobutton-stub
        modelvalue="1"
        name="answer"
        value="2"
      />
      
      a3
    
    </label>
  </p>
</div>
`;
